java/statestore/src/main/java/sleeper/statestore/dynamodb/DynamoDBFileReferenceStore.java

```diff
@@ -115,22 +127,46 @@ public void addFilesWithReferences(List<AllReferencesToAFile> files) throws Stat
             addFileReferenceCount(file.getFilename(), file.getExternalReferenceCount(), updateTime);
             for (FileReference reference : file.getInternalReferences()) {
-                addFile(reference, updateTime);
+                addFileReference(reference, updateTime);
             }
         }
     }
 
+    private void addFileReference(FileReference fileReference, Instant updateTime) throws StateStoreException {
+        try {
+            TransactWriteItemsResult transactWriteItemsResult = dynamoDB.transactWriteItems(new TransactWriteItemsRequest()
+                    .withTransactItems(
+                            new TransactWriteItem().withPut(putNewFile(fileReference, updateTime)),
+                            new TransactWriteItem().withUpdate(fileReferenceCountUpdate(fileReference.getFilename(), updateTime, 1)))
+                    .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL));
+            List<ConsumedCapacity> consumedCapacity = transactWriteItemsResult.getConsumedCapacity();
+            double totalConsumed = consumedCapacity.stream().mapToDouble(ConsumedCapacity::getCapacityUnits).sum();
+            LOGGER.debug("Put file reference for file {} to table {}, read capacity consumed = {}",
+                    fileReference.getFilename(), activeTableName, totalConsumed);
+        } catch (TransactionCanceledException e) {
+            if (hasConditionalCheckFailure(e)) {
+                throw new FileReferenceAlreadyExistsException(fileReference);
+            }
+        } catch (AmazonDynamoDBException e) {
+            throw new StateStoreException("Failed to add file", e);
+        }
+    }
+
     private void addFile(FileReference fileReference, Instant updateTime) throws StateStoreException {
         try {
             TransactWriteItemsResult transactWriteItemsResult = dynamoDB.transactWriteItems(new TransactWriteItemsRequest()
                     .withTransactItems(
                             new TransactWriteItem().withPut(putNewFile(fileReference, updateTime)),
-                            new TransactWriteItem().withUpdate(fileReferenceCountUpdateAddingFile(fileReference, updateTime)))
+                            new TransactWriteItem().withPut(fileReferenceCountPut(fileReference.getFilename(), updateTime)))
                     .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL));
             List<ConsumedCapacity> consumedCapacity = transactWriteItemsResult.getConsumedCapacity();
             double totalConsumed = consumedCapacity.stream().mapToDouble(ConsumedCapacity::getCapacityUnits).sum();
             LOGGER.debug("Put file reference for file {} to table {}, read capacity consumed = {}",
                     fileReference.getFilename(), activeTableName, totalConsumed);
+        } catch (TransactionCanceledException e) {
+            if (hasConditionalCheckFailure(e)) {
+                throw new FileAlreadyExistsException(fileReference.getFilename(), e);
+            }
```

Comment:
This ignores the failure entirely if it's not a conditional check failure. That case needs to be rethrown as a StateStoreException.